File: Week_02/264/ugly_number_ii2.py
class Solution:
    def nthUglyNumber(self, n: int) -> int:
        dp = [0] * n
        dp[0] = 1

        two = three = five = 0

        for i in range(1, n):
            num = min(2 * dp[two], 3 * dp[three], 5 * dp[five])
            dp[i] = num
            if 2 * dp[two] == num:
                two += 1
            if 3 * dp[three] == num:
                three += 1
            if 5 * dp[five] == num:
                five += 1
        return dp[-1]

    def test(self):
        for n, target in [
            (1, 1),
            (10, 12),
        ]:
            ans = self.nthUglyNumber(n)
            assert ans == target, f"n: {n}, target: {target}"
        print("well done")


if __name__ == "__main__":
    Solution().test()

File: Week_02/264/ugly_number_ii2.go
package main

func nthUglyNumber2(n int) int {
	dp := make([]int, n)
	dp[0]=1
	two := 0
	three := 0
	five := 0
	for i := 1; i < n; i++ {
		num := min(2*dp[two], 3*dp[three], 5*dp[five])
		dp[i] = num
		if 2*dp[two] == num {
			two++
		}
		if 3*dp[three] == num {
			three++
		}
		if 5*dp[five] == num {
			five++
		}
	}
	return dp[n-1]
}

func min(x, y, z int) int {
	if x <= y && x <= z {
		return x
	}
	if y <= x && y <= z {
		return y
	}
	return z
}
